state.go

```diff
@@ -1332,9 +1346,42 @@ func LoadStateData(store store.Store) (StateData, error) {
 	}
 }
 
+func LoadStateData(store store.Store) (StateData, error) {
+	_, err := store.ReadAll(stateDataKeyCustom)
+	if err == os.ErrNotExist {
+		// we don't have a custom data which means
+		// we are reading the state data
+		return loadData(store, stateDataKey)
+	} else if err != nil {
+		return StateData{}, err
+	}
+
+	// we have custom data
+	data, err := loadData(store, stateDataKey)
+	if err == os.ErrNotExist {
+		return loadData(store, stateDataKeyCustom)
+	} else if err != nil {
+		return StateData{}, err
+	}
+	key := data.UpdateInfo.ID
+
+	customData, err := loadData(store, stateDataKeyCustom)
+	if err != nil {
+		return StateData{}, err
+	}
+
+	if key == customData.UpdateInfo.ID {
+		return customData, nil
+	}
+	return data, nil
```

Comment:
Once we have proper transactions I think these reads should be done in one transaction as well.

Reply:
I had to rework this quite a bit to fit into my branch, and while doing so I discovered an additional issue that we should handle: This method only works once. If we upgrade the database schema again, then we can't reuse this method. The reason is that the key lies in the old client not using the uncommitted key, and the new client using it. But this doesn't work if the client always uses it. We would have to invent an additional "superuncommitted" key!

I solved this in the following way: I changed the uncommitted key to only be used when the original key has a version that is different from the version used by the current client. This always keeps the old client using the normal key, and the new client, once it discovers the normal key has an older version, uses the uncommitted key. You can check out the work [here](https://github.com/kacf/mender/commit/f630e73981397dd8808c12761b99a01473b7d906).

Reply:
The rest of the logic regarding the `UpdateInfo.ID` is the same.